ext/random/randomizer.c

```diff
@@ -0,0 +1,285 @@
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Go Kudo <g-kudo@colopl.co.jp>                                |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "php_random.h"
+
+#include "ext/standard/php_array.h"
+#include "ext/standard/php_string.h"
+
+#include "ext/spl/spl_exceptions.h"
+#include "Zend/zend_exceptions.h"
+
+static inline void randomizer_common_init(php_random_randomizer *randomizer, zend_object *engine_object) {
+	if (engine_object->ce->type == ZEND_INTERNAL_CLASS) {
+		/* Internal classes always php_random_engine struct */
+		php_random_engine *engine = php_random_engine_from_obj(engine_object);
+		
+		/* Copy engine pointers */
+		randomizer->algo = engine->algo;
+		randomizer->status = engine->status;
+	} else {
+		/* Self allocation */
+		randomizer->status = php_random_status_alloc(&php_random_algo_user, false);
+		php_random_status_state_user *state = randomizer->status->state;
+		zend_string *mname;
+		zend_function *generate_method;
+
+		mname = zend_string_init("generate", sizeof("generate") - 1, 0);
+		generate_method = zend_hash_find_ptr(&engine_object->ce->function_table, mname);
+		zend_string_release(mname);
+
+		/* Create compatible state */
+		state->object = engine_object;
+		state->generate_method = generate_method;
+
+		/* Copy common pointers */
+		randomizer->algo = &php_random_algo_user;
+
+		/* Mark self-allocated for memory management */
+		randomizer->is_userland_algo = true;
+	}
+}
+
+/* {{{ Random\Randomizer::__construct() */
+PHP_METHOD(Random_Randomizer, __construct)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_object *engine_object = NULL;
+	zval zengine_object;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS_OR_NULL(engine_object, random_ce_Random_Engine);
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* Create default RNG instance */
+	if (!engine_object) {
+		engine_object = random_ce_Random_Engine_Secure->create_object(random_ce_Random_Engine_Secure);
+
+		/* No need self-refcount */
+		GC_DELREF(engine_object);
+	}
+
+	ZVAL_OBJ(&zengine_object, engine_object);
+
+	zend_update_property(random_ce_Random_Randomizer, Z_OBJ_P(ZEND_THIS), "engine", strlen("engine"), &zengine_object);
```

Comment:
First, `strlen("engine")` could be "optimized" to `sizeof("engine") - 1`

then, `"engine", sizeof("engine") - 1` could be "deduplicated" to `ZEND_STRL("engine")`
and similarly in https://github.com/php/php-src/blob/9490d0c32ea3c5e62f42218cf90a6b1d5fbfa33a/ext/random/engine_mt19937.c#L397, https://github.com/php/php-src/blob/9490d0c32ea3c5e62f42218cf90a6b1d5fbfa33a/ext/random/randomizer.c#L45 and https://github.com/php/php-src/blob/9490d0c32ea3c5e62f42218cf90a6b1d5fbfa33a/ext/random/randomizer.c#L277

(as well as https://github.com/php/php-src/blob/26d890e6ba7963c4750ab102b3c64b86a6574b51/Zend/zend_attributes.c#L117 @TimWolla)

Reply:
> First, strlen("engine") could be "optimized" to sizeof("engine") - 1

Modern compilers are perfectly capable of understanding `strlen()` for a constant string and optimizing that to a constant.

> could be "deduplicated" to ZEND_STRL("engine") and similarly in 

This is possible indeed. But it's exceedingly rarely used. I believe it would be more useful to replace all of those with a single sweeping change for consistency.

Reply:
It might be best to avoid the use of `ZEND_STRL()` altogether, since it may not play well in combination with other macros.

Reply:
As noted `strlen` is optimised by most compilers so should be fine. I don't see this comment like something that should hold this PR though.

Reply:
fixed: f73d51a0695b87e0ddadd6606432e8af81a83729

Reply:
@zeriyoshi you did the exact *opposite* of what I was suggesting... Whatever, at least now it's consistent 😆 (and as I said, it was a "nit", never intended to "hold this PR")

Reply:
@guilliamxavier 
Oops! I chose this one based on its usage in the entire source code, but it certainly may not be modern...
Will correct during follow-up PR.

Reply:
Forget "modernity" 😄 I first invoked "optimization" but apparently it's already good, then "deduplication" but apparently it may have bad effects, so remaining the most important is "consistency": you had 3 occurrences of `sizeof("...") - 1` and 1 of `strlen("...")`, I suggested to change this 1 like those 3, you did the reverse (i.e. changed those 3 like this 1) but that's OK too :wink: For your follow-up PR you can/should focus on "just" splitting the overloaded getInt(), I think